src/datepicker/utils/date-helpers.js

```diff
@@ -25,6 +48,263 @@ class DateHelpers<T> {
+    return (
+      (!!minDate &&
+        this.differenceInCalendarMonths(minDate, previousMonth) > 0) ||
+      (!!includeDates &&
+        includeDates.every(
+          includeDate =>
+            this.differenceInCalendarMonths(includeDate, previousMonth) > 0,
+        )) ||
+      false
+    );
+  };
+  monthDisabledAfter: (T, {maxDate: ?T, includeDates: ?Array<T>}) => boolean = (
+    day,
+    {maxDate, includeDates} = {},
+  ) => {
+    const nextMonth = this.adapter.addMonths(day, 1);
+    return (
+      (!!maxDate && this.differenceInCalendarMonths(nextMonth, maxDate) > 0) ||
+      (!!includeDates &&
+        includeDates.every(
+          includeDate =>
+            this.differenceInCalendarMonths(nextMonth, includeDate) > 0,
+        )) ||
+      false
+    );
+  };
+  setDate: (T, number) => T = (date, dayNumber) => {
+    const startOfMonthNoTime = this.adapter.startOfMonth(date);
+    const startOfMonthHoursAndMinutes = this.adapter.mergeDateAndTime(
+      startOfMonthNoTime,
+      date,
+    );
+    const startOfMonth = this.adapter.setSeconds(
+      startOfMonthHoursAndMinutes,
+      this.adapter.getSeconds(date),
+    );
+    return this.adapter.addDays(startOfMonth, dayNumber - 1);
+  };
+  getDate: T => number = date => Number(this.adapter.formatByString(date, 'd'));
+  applyDateToTime: (?T, T) => T = (time, date) => {
+    if (!time) return date;
+    const yearNumber = this.adapter.getYear(date);
+    const monthNumber = this.adapter.getMonth(date);
+    const dayNumber = this.getDate(date);
+    const yearDate = this.adapter.setYear(time, yearNumber);
+    const monthDate = this.adapter.setMonth(yearDate, monthNumber);
+    return this.setDate(monthDate, dayNumber);
+  };
+  applyTimeToDate: (?T, T) => T = (date, time) => {
+    if (!date) return time;
+    return this.adapter.setSeconds(
+      this.adapter.mergeDateAndTime(date, time),
+      0,
+    );
+  };
+  isDayDisabled: (
+    T,
+    {
+      minDate: ?T,
+      maxDate: ?T,
+      excludeDates: ?Array<T>,
+      includeDates: ?Array<T>,
+      filterDate: ?(day: T) => boolean,
+    },
+  ) => boolean = (
+    day,
+    {minDate, maxDate, excludeDates, includeDates, filterDate} = {},
+  ) => {
+    return (
+      this.isOutOfBounds(day, {minDate, maxDate}) ||
+      (excludeDates &&
+        excludeDates.some(excludeDate =>
+          this.adapter.isSameDay(day, excludeDate),
+        )) ||
+      (includeDates &&
+        !includeDates.some(includeDate =>
+          this.adapter.isSameDay(day, includeDate),
+        )) ||
+      (filterDate && !filterDate(day)) ||
+      false
+    );
+  };
+  isOutOfBounds: (T, {minDate: ?T, maxDate: ?T}) => boolean = (
+    day,
+    {minDate, maxDate} = {},
+  ) => {
+    return (
+      (!!minDate && this.differenceInCalendarDays(day, minDate) < 0) ||
+      (!!maxDate && this.differenceInCalendarDays(day, maxDate) > 0)
+    );
+  };
+  setSeconds: (T, number) => T = (date, seconds) =>
```

Comment:
I don't love having to define all of these functions manually even though they're just the same calls with the exact same arguments in `date-io` the problem was, this was the only way I could find where those functions would be correctly typed.  Every thing else dynamic function creation and that caused flow to break down.